File: ArchitectureModel/src/main/java/com/kailin/architecture_model/encrypt/AESUtil.java
package com.kailin.architecture_model.encrypt;

import android.security.keystore.KeyProperties;

import javax.crypto.spec.IvParameterSpec;

public class AESUtil extends CipherUtil {

    private static volatile AESUtil util;

    public static AESUtil getInstance() {
        if (util == null) {
            synchronized (CipherUtil.class) {
                if (util == null)
                    util = new AESUtil();
            }
        }
        return util;
    }

    public static AESUtil createInstance(String keyString, String ivParameterSpec) {
        synchronized (CipherUtil.class) {
            util = new AESUtil(keyString, ivParameterSpec);
        }
        return util;
    }

    protected String keyString;
    protected IvParameterSpec ivParameterSpec;
    protected final String algorithm = KeyProperties.KEY_ALGORITHM_AES;
    protected final String transformation = String.format(ALGORITHM_FORMAT,
            KeyProperties.KEY_ALGORITHM_AES,
            KeyProperties.BLOCK_MODE_CBC,
            "PKCS5Padding");

    protected AESUtil() {
        keyString = "kailin12kailin12kailin12";
        ivParameterSpec = new IvParameterSpec("kailin12kailin12".getBytes());
    }

    protected AESUtil(String keyString, String ivParameterSpec) {
        this.keyString = keyString;
        this.ivParameterSpec = new IvParameterSpec(ivParameterSpec.getBytes());
    }

    @Override
    protected String getKeyString() {
        return keyString;
    }

    @Override
    protected IvParameterSpec getIvParameterSpec() {
        return ivParameterSpec;
    }

    @Override
    protected String getAlgorithm() {
        return algorithm;
    }

    @Override
    protected String getTransformation() {
        return transformation;
    }
}

File: ArchitectureModel/src/main/java/com/kailin/architecture_model/util/security/AESUtil.kt
package com.kailin.utillibrary.security

import android.security.keystore.KeyProperties
import java.security.Key
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicReference
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class AESUtil private constructor(private val keyString: String, private val ivString: String) : CipherUtil() {

    override val key: Key
        get() = SecretKeySpec(keyString.toByteArray(), KeyProperties.KEY_ALGORITHM_AES)

    override val ivParameterSpec: IvParameterSpec
        get() = IvParameterSpec(ivString.toByteArray())

    override val transformation: String
        get() = "${KeyProperties.KEY_ALGORITHM_AES}/${KeyProperties.BLOCK_MODE_CBC}/PKCS5Padding"

    companion object {
        private val referenceMap = ConcurrentHashMap<String, AtomicReference<AESUtil>>()

         fun getInstance(keyString: String, ivString: String): AESUtil {
            var reference = referenceMap[keyString]
            if (reference == null) {
                reference = AtomicReference()
                referenceMap[keyString] = reference
            }

            while (true) {
                var util = reference.get()
                if (util != null) {
                    return util
                }

                util = AESUtil(keyString, ivString)
                if (reference.compareAndSet(null, util)){
                    return util
                }
            }
        }
    }
}